platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/evaluator/conventions/SoftwareTypeConventionHandler.kt

```diff
@@ -0,0 +1,131 @@
+import org.gradle.internal.declarativedsl.analysis.AssignmentRecord
+import org.gradle.internal.declarativedsl.analysis.ObjectOrigin
+import org.gradle.internal.declarativedsl.analysis.ResolutionResult
+import org.gradle.internal.declarativedsl.analysis.ResolutionTrace
+import org.gradle.internal.declarativedsl.conventions.softwareTypeRegistryBasedConventionRepository
+import org.gradle.internal.declarativedsl.evaluator.DeclarativeDslNotEvaluatedException
+import org.gradle.internal.declarativedsl.evaluator.conversion.AnalysisAndConversionStepRunner
+import org.gradle.internal.declarativedsl.evaluator.conversion.ConversionStepContext
+import org.gradle.internal.declarativedsl.evaluator.runner.AbstractAnalysisStepRunner
+import org.gradle.internal.declarativedsl.evaluator.runner.AnalysisStepContext
+import org.gradle.internal.declarativedsl.evaluator.runner.EvaluationResult.Evaluated
+import org.gradle.internal.declarativedsl.evaluator.runner.EvaluationResult.NotEvaluated
+import org.gradle.internal.declarativedsl.evaluator.runner.ParseAndResolveResult
+import org.gradle.internal.declarativedsl.language.Assignment
+import org.gradle.internal.declarativedsl.language.Block
+import org.gradle.internal.declarativedsl.language.Expr
+import org.gradle.internal.declarativedsl.language.LanguageTreeResult
+import org.gradle.internal.declarativedsl.language.SourceData
+import org.gradle.internal.declarativedsl.language.SourceIdentifier
+import org.gradle.internal.declarativedsl.project.projectInterpretationSequenceStep
+import org.gradle.plugin.software.internal.ConventionHandler
+import org.gradle.plugin.software.internal.SoftwareTypeRegistry
+
+
+/**
+ * A {@link ConventionHandler} for applying declarative conventions.
+ */
+class SoftwareTypeConventionHandler(softwareTypeRegistry: SoftwareTypeRegistry) : ConventionHandler {
+    private
+    val step = projectInterpretationSequenceStep(softwareTypeRegistry)
+    private
+    val conventionRepository = softwareTypeRegistryBasedConventionRepository(softwareTypeRegistry)
+
+    override fun apply(target: Any, softwareTypeName: String) {
+        val analysisStepRunner = ApplyConventionsOnlyAnalysisStepRunner()
+        val analysisStepContext = AnalysisStepContext(
+            emptySet(),
+            setOf(
+                ConventionApplicationHandler { listOf(conventionRepository.findConventions(softwareTypeName)).requireNoNulls() }
+            )
+        )
+
+        val result = AnalysisAndConversionStepRunner(analysisStepRunner)
+            .runInterpretationSequenceStep(
+                "<none>",
+                "",
+                step,
+                ConversionStepContext(target, analysisStepContext)
+            )
+
+        when (result) {
+            is Evaluated -> Unit
+            is NotEvaluated -> {
+                // TODO: What kind of errors can actually come out of the conversion step?  Should we have a better exception type since we're
```

Comment:
Ok, I've updated the comment to state that we shouldn't expect anything here, but I left the exception in place in case something changes in the future.